win32ss/user/user32/windows/cursoricon.c

```diff
@@ -7,11 +7,376 @@
+        width, height, bit_depth, color_type, interlace_type);
+
+    channels = png_get_channels(png_ptr, info_ptr);
+    TRACE("channels is %d\n", channels);
+
+    // row_pointers
+    row_pointers = png_get_rows(png_ptr, info_ptr);
+    width1 = png_get_image_width(png_ptr, info_ptr);
+    height1 = png_get_image_height(png_ptr, info_ptr);
+
+    size = width1 * channels;
+
+    TRACE("size %d, width1 %d, height1 %d\n",
+        size, width1, height1);
+    rowbytes = (int)png_get_rowbytes(png_ptr, info_ptr); // same as size above
+    image_size = height * rowbytes;
+
+    // Read png image data
+    // Set row pointer which will take pixel value from png file
+    row_pointers = (png_bytepp)png_malloc(png_ptr, sizeof(png_bytepp) * height);
+
+    for (int i = 0; i < height; i++)
+    {
+        row_pointers[i] = png_malloc(png_ptr, png_get_rowbytes(png_ptr, info_ptr));
+    }
+
+    // Set row pointer to the png struct
+    png_set_rows(png_ptr, info_ptr, row_pointers);
+
+    // Read png image data and save in row pointer
+    // After reading the image, you can deal with the image data with row pointers
+    png_read_image(png_ptr, row_pointers);
+
+    png_read_end(png_ptr, info_ptr);
+
+    data = NULL;
+    size += size % 4; // Align
+    size *= height;
+    data = malloc(size);
+    pos = 0;
+    stride=channels;
+
+    for(int i = height-1; i >= 0; i--)
+    {
+        for(int j = 0; j <stride*width; j += stride)
+        {
+            data[pos++] = row_pointers[i][j+2]; // red
+            data[pos++] = row_pointers[i][j+1]; // green
+            data[pos++] = row_pointers[i][j];  // blue
+            if(stride==4)
+            {
+                data[pos++] = row_pointers[i][j+3]; // alpha
+            }
+        }
+        pos+=(stride*width) % 4;
+    }
+
+    if (FindTempDirectoryA() == NULL)
+        ERR("Temp Directory Not Found\n");
+    else
+    {
+        strcpy(lpTempPathBuffer, FindTempDirectoryA());
+        TRACE("Temp File Name is %s\n", lpTempPathBuffer);
+    }
+
+    /* Clean up after the read, and free any memory allocated */
+    png_destroy_read_struct(&png_ptr, &info_ptr, png_infopp_NULL);
+
+    bpp = 0;
+
+    switch (color_type)
+    {
+    case PNG_COLOR_TYPE_RGB:
+        if (bit_depth == 8)
+            bpp = 24;
+        break;
+
+    case PNG_COLOR_TYPE_RGB_ALPHA:
+        if (bit_depth == 8)
+        {
+            png_set_bgr(png_ptr);
+            bpp = 32;
+        }
+        break;
+
+    default:
+        break;
+    }
+
+    if (!bpp)
+    {
+        FIXME("unsupported PNG color format %d, %d bpp\n", color_type, bit_depth);
+        png_destroy_read_struct(&png_ptr, &info_ptr, png_infopp_NULL);
+        return;
+    }
+
+    remove(lpTempPathBuffer);
+    fp = fopen(lpTempPathBuffer, "ab");
+    if (!fp)
+    {
+        ERR("File Open Failed for '%s'.\n", lpTempPathBuffer);
+    }
+
+    if (fp == INVALID_HANDLE_VALUE)
```

Comment:
```suggestion
    if (fp == NULL)
```

Reply:
We check for NULL is line 333. This is to test for other possible problems. Thanks.